File: examples/node_js/template_index.js
const facerec = require('face_sdk_3divi');
const yargs = require('yargs');
const fs = require('fs');
const path = require('path');

function parse_args()
{
	return yargs
		.version(false)
		.usage('Usage: $0 --input_image <path> [options]')
		.option('input_image', {type: 'string', describe: 'Path to input image', demandOption: true})
		.option('images_path_dir', {type: 'string', describe: 'Path to dir with images', demandOption: true})
		.option('sdk_path', {type: 'string', default: '../../', describe: 'Path to the SDK'})
		.option('use_cuda', {type: 'boolean', default: false, describe: 'Use CUDA acceleration',})
		.option('modification', {type: 'string', default: '100', describe: 'Modification string',})
		.help()
		.argv
}


function createTemplateIndex(imageDirPath, modification, pipeline)
{
	const index_config = new facerec.Context();
	index_config.get("modification").value = modification
	index_config.get("version").value = 1
	let template_index = new facerec.DynamicTemplateIndex(index_config);

	const files = fs.readdirSync(imageDirPath);

	const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'];
	const imageFiles = files.filter(file => {
		const ext = path.extname(file).toLowerCase();
		return imageExtensions.includes(ext);
	});

	imageFiles.forEach(image_name => {
		const data = fs.readFileSync(`${imageDirPath}/${image_name}`);
		let ioData = new facerec.Context(data);

		pipeline.forEach(pb => {
			pb.process(ioData);
		});

		template_index.add(
			ioData.get("objects").get(0).get("face_template").get("template").value,
			image_name
		);
	});

	return template_index;
}

function main()
{
	try {
		const argv = parse_args();
		const sdk_path = argv.sdk_path;
		const dll_path = sdk_path + "/lib/libfacerec.so";
		const conf_path = sdk_path + "/conf/facerec";
		const license_path = sdk_path + "/license";

		facerec.Init(dll_path, conf_path, license_path);

		const input_image = argv.input_image;
		const modification = argv.modification;
		const use_cuda = argv.use_cuda;

		fontScale = argv.font_scale;

		let detector_config = new facerec.Context();
		detector_config.get("unit_type").value = "FACE_DETECTOR";
		detector_config.get("modification").value = "ssyv_light";
		detector_config.get("use_cuda").value = use_cuda;

		let fitter_config = new facerec.Context();
		fitter_config.get("unit_type").value = "FACE_FITTER";
		fitter_config.get("use_cuda").value = use_cuda;

		let extractor_config = new facerec.Context();
		extractor_config.get("unit_type").value = "FACE_TEMPLATE_EXTRACTOR";
		extractor_config.get("use_cuda").value = use_cuda;
		extractor_config.get("modification").value = modification;

		const pipeline = [
			new facerec.ProcessingBlock(detector_config),
			new facerec.ProcessingBlock(fitter_config),
			new facerec.ProcessingBlock(extractor_config),
		];

		const index_config = new facerec.Context();
		index_config.get("modification").value = modification
		index_config.get("version").value = 1
		const template_index = new facerec.DynamicTemplateIndex(index_config);

		const images_path_dir = argv.images_path_dir;

		template_index.concat(createTemplateIndex(images_path_dir, modification, pipeline));


		const data = fs.readFileSync(input_image);
		let ioData = new facerec.Context(data);
		pipeline.forEach(pb => { pb.process(ioData); });
		for (const pb of pipeline)
		{
			pb.process(ioData);
		}

		const matcher_config = new facerec.Context();
		matcher_config.get("unit_type").value = "MATCHER_MODULE";
		matcher_config.get("modification").value = modification;
		const matcher = new facerec.ProcessingBlock(matcher_config);

		let matcherData = new facerec.Context();
		matcherData.get("template_index").value = template_index;
		matcherData.get("queries").value = ioData.get("objects").get(0).get("face_template").get("template");

		matcher.process(matcherData);

		const result = matcherData.get("results").get(0);
		console.log(`UUID: ${result.get("uuid").value}`)
		console.log(`Score: ${result.get("score").value}`)
	} catch (err) {
		console.log(err)
	}
}

main();

File: examples/python/processing_blocks/template_index.py
import argparse  # for an ArgumentParser
import json
import os.path
import imghdr
import shutil
from io import BytesIO
from typing import List

import sys
try:
    from face_sdk_3divi import FacerecService  # FacerecService creates service
    from face_sdk_3divi.modules.context_template import ContextTemplate
    from face_sdk_3divi.modules.processing_block import ProcessingBlock
    from face_sdk_3divi.modules.dynamic_template_index import DynamicTemplateIndex
except:
    from face_sdk_3divi import FacerecService, ContextTemplate, ProcessingBlock, DynamicTemplateIndex # FacerecService creates service
    

def help_message():
    message = f"\n This program is an example of the TemplateIndex integration \n Usage: " \
              " [--sdk_path <sdk_root_dir>] " \
              " [--use_cuda <--use_cuda>]" \
              " [--modification <modification>]" \
              " [--input_image <input_image>]" \
              " [--capacity <capacity>]" \
              " [--templates_description <templates_description>]" \
              " [--images_paths <images_path images_path...>]" \
              " [--templates_paths <templates_path templates_path...>]"

    print(message)


def parse_args():  # launch parameters
    parser = argparse.ArgumentParser(description="Template Index Example")
    parser.add_argument("--sdk_path", default="../../../", type=str)
    parser.add_argument("--use_cuda",
                        action="store_true")  # pass the '--use_cuda' parameter before launch to use cuda acceleration
    parser.add_argument("--input_image", type=str, required=True)
    parser.add_argument("--capacity", type=int, default=1000)
    parser.add_argument("--modification", type=str, default="1000")
    parser.add_argument("--images_paths", nargs="+", type=str, default=["../../../bin/set1"])
    parser.add_argument("--templates_paths", nargs="+", type=str)
    parser.add_argument("--templates_description", type=str)

    return parser.parse_args()


def read_image(input_image: str) -> bytes:
    with open(input_image, "rb") as file:
        return file.read()


def create_service(sdk_path: str) -> FacerecService:
    sdk_conf_dir = os.path.join(sdk_path, "conf", "facerec")
    if sys.platform == "win32":  # for Windows
        sdk_dll_path = os.path.join(sdk_path, "bin", "facerec.dll")
    else:  # for Linux
        sdk_dll_path = os.path.join(sdk_path, "lib", "libfacerec.so")

    return FacerecService.create_service(  # create FacerecService
        sdk_dll_path,
        sdk_conf_dir,
        f'{sdk_path}/license')


def create_detector(use_cuda: bool) -> ProcessingBlock:
    return service.create_processing_block({
        "unit_type": "FACE_DETECTOR",
        "modification": "ssyv_light",
        "use_cuda": use_cuda,
    })


def create_fitter(use_cuda: bool) -> ProcessingBlock:
    return service.create_processing_block({
        "unit_type": "FACE_FITTER",
        "modification": "fda",
        "use_cuda": use_cuda,
    })


def create_template_extractor(modification: str, use_cuda: bool):
    return service.create_processing_block({
        "unit_type": "FACE_TEMPLATE_EXTRACTOR",
        "modification": modification,
        "use_cuda": use_cuda,
    })


def is_image(file_path):
    try:
        return imghdr.what(file_path) is not None
    except FileNotFoundError:
        return False


def create_template_index(images_path: str, modification: str, pipeline: List[ProcessingBlock]):

    config = service.create_context({"modification": modification, "version": 1})
    template_index = service.create_dynamic_template_index(config)

    for root, dirs, files in os.walk(images_path):
        for file in files:
            file_path = os.path.join(root, file)

            if not is_image(file_path):
                continue

            data = service.create_context_from_encoded_image(read_image(file_path))

            for processing_block in pipeline:
                processing_block(data)

            if not data.contains("objects"):
                continue

            template_index.add(
                data["objects"][0]["face_template"]["template"].get_value(),
                file_path
            )

    return template_index


def load_templates(templates_path: str, template_index: DynamicTemplateIndex, templates_description: str):
    file_size = os.path.getsize(templates_path)
    templates: List[ContextTemplate] = []
    uuids: List[str] = []
    description_file_name = os.path.basename(templates_path)

    with open(templates_description, "r") as file:
        description = json.loads(file.read())

    with open(templates_path, "rb") as file:
        templates_data = BytesIO()

        shutil.copyfileobj(file, templates_data)

        templates_data.seek(0)

        while True:
            offset = templates_data.tell()

            if offset == file_size:
                break

            templates.append(service.load_context_template(templates_data))

            if description_file_name in description:
                uuids.append(f"Image {description[description_file_name][str(offset)]} from templates file: {templates_path}")
            else:
                uuids.append(f"{templates_path}_{offset}")

            if len(templates) == 100:
                template_index.add(templates, uuids)

                templates.clear()
                uuids.clear()

    if len(templates) > 0:
        template_index.add(templates, uuids)

        templates.clear()
        uuids.clear()


if __name__ == "__main__":
    help_message()
    args = parse_args()

    if not (args.images_paths or args.templates_paths):
        args.error("At least one of --images_paths or --templates_paths is required.")

    service = create_service(args.sdk_path)

    if sys.platform == "win32":  # for Windows
        sdk_onnx_path = os.path.join(args.sdk_path, "bin")
    else:  # for Linux
        sdk_onnx_path = os.path.join(args.sdk_path, "lib")

    # Create processing pipeline: detect face -> construct face keypoints -> extract face template
    pipeline: List[ProcessingBlock] = [
        create_detector(args.use_cuda),
        create_fitter(args.use_cuda),
        create_template_extractor(args.modification, args.use_cuda)
    ]

    config = service.create_context({
        "capacity": args.capacity,
        "modification": args.modification,
        "version": 1,
    })
    template_index = service.create_dynamic_template_index(config)

    if args.images_paths is not None:
        print("Images processing...")
        for images_path in args.images_paths:
            template_index.concat(create_template_index(images_path, args.modification, pipeline))

    if args.templates_paths is not None:
        print("Templates loading....")
        for templates_path in args.templates_paths:
            load_templates(templates_path, template_index, args.templates_description)

    data = service.create_context_from_encoded_image(read_image(args.input_image))

    print(f"Process {args.input_image}")

    for processing_block in pipeline:
        processing_block(data)

    if not data.contains("objects"):
        print(f"No face in {args.input_image}")

        sys.exit(1)

    matcher_module = service.create_processing_block({
        "unit_type": "MATCHER_MODULE",
        "modification": args.modification
    })
    matcher_data = service.create_context({
        "template_index": template_index,
        "queries": data["objects"][0]["face_template"]["template"],
    })

    print("Matching")

    matcher_module(matcher_data)

    result = matcher_data["results"][0]

    print(f"UUID: {result['uuid'].get_value()}")
    print(f"Score: {result['score'].get_value()}")
